### ISSUE TITLE
[BUG] Type variables are not instantiated during type-checking 

### ISSUE THREAD

**Issue by andrey-kuprianov** [informalsystems/apalache#869]
## Description

Snowcat doesn't seem to instantiate type variables when doing type-checking; as a result polymorphic type signatures do not work:

```sh
Type checking snowcat_bug.tla                                     I@09:04:16.784
PASS #0: SanyParser                                               I@09:04:17.038
Parsing file /root/agoric/snowcat_bug.tla
PASS #1: TypeCheckerSnowcat                                       I@09:04:17.532
 > Running Snowcat .::.                                           I@09:04:17.533
[snowcat_bug.tla:23:18-23:41]: No match between operator signature (((Str -> Int), Str, Int) => (Str -> Int)) and arguments (Int -> Str) and Int and Str E@09:04:17.745
[snowcat_bug.tla:21:1-23:41]: Error when computing the type of Next E@09:04:17.748
 > Snowcat asks you to fix the types. Meow.                       I@09:04:17.749
Type checker [FAILED]                                             I@09:04:17.749
It took me 0 days  0 hours  0 min  1 sec                          I@09:04:17.750
Total time: 1.37 sec                                              I@09:04:17.750
EXITCODE: ERROR (255)
```

## Input specification

Here is the MWE:

```tla
------------------------------------ MODULE snowcat_bug -----------------------------

\* Extend a function with a key -> value mapping
\* @type: (a -> b, a, b) => a -> b;
Extend(function, key, value) == 
    [k \in DOMAIN function \union {key} |-> IF k = key THEN value ELSE function[k]]


VARIABLE 
\* @type: Str -> Int;
strToInt,
\* @type: Int -> Str;
intToStr


Init == 
  /\ strToInt = [ s \in {} |-> 0 ]
  /\ intToStr = [ i \in {} |-> "" ]


Next ==
  /\ strToInt' = Extend(strToInt, "a", 1)
  /\ intToStr' = Extend(intToStr, 1, "a")
    
===============================================================================
```

## The command line parameters used to run the tool

```sh
apalache typecheck snowcat_bug.tla
```

## System information

- Apalache version : 0.15.7 build e6eda01
- OS: Ubuntu 20.04.2 LTS
- JDK version: openjdk version "15.0.2" 2021-01-19

**Comment by konnov** [comment]
I think we missed instantiation of type variables somewhere. This should be fixable.

**Comment by konnov** [comment]
I have been working on a systematic solution to this issue for a few days, and it is still breaking here and there. So it will take longer than anticipated. The problem is that inference of parametric operators was a by-product of the approach, not the primary goal. @andrey-kuprianov, how quick do you want to have it?

**Comment by andrey-kuprianov** [comment]
don't worry -- it's more a nice-to-have, not critical at all

**Comment by konnov** [comment]
Ok. It requires some careful thinking, so we do not introduce bugs in a quick fix

**Comment by konnov** [comment]
This bug is partially fixed in #952. The second half of the issue should be fixed in #953.

**Comment by konnov** [comment]
Fixed via #952
